src/Dscp/DB/SQLite/Instances.hs

```diff
@@ -1,28 +1,305 @@
+    fromRow = Assignment <$> field <*> field <*> field
+
+instance ToRow Assignment where
+    toRow task@ (Assignment course ty text) =
+        [toField (hash task), toField course, toField ty, toField text]
+
+instance FromRow Submission where
+    fromRow = Submission <$> field <*> field <*> fromRow
+
+-- These three structures depend on the public key, which is not in db.
+-- Therefore, we have to plug it in somehow.
+instance FromRow (PublicKey -> SubmissionWitness) where
+    fromRow = (\sig pk -> SubmissionWitness pk sig) <$> field
+
+instance FromRow (PublicKey -> SignedSubmission) where
+    fromRow = (\sub wit pk -> SignedSubmission sub (wit pk)) <$> fromRow <*> fromRow
+
+instance FromRow (PublicKey -> PrivateTx) where
+    fromRow = (\sigSub grade time pk -> PrivateTx (sigSub pk) grade time) <$> fromRow <*> field <*> field
+
+getStudentCourses :: DBM m => StudentId -> m [CourseId]
+getStudentCourses student =
+    query getStudentCoursesQuery (Only student)
+  where
+    getStudentCoursesQuery :: Query
+    getStudentCoursesQuery = [q|
+        select  course_id
+        from    StudentCourses
+        where   student_addr = ?
+    |]
+
+enrollStudentToCourse :: DBM m => StudentId -> CourseId -> m ()
+enrollStudentToCourse student course = do
+    execute enrollStudentToCourseRequest (student, course)
+  where
+    enrollStudentToCourseRequest :: Query
+    enrollStudentToCourseRequest = [q|
+        insert into  StudentCourses
+        values       (?, ?)
+    |]
+
+getStudentAssignments :: DBM m => StudentId -> CourseId -> m [Assignment]
+getStudentAssignments student course = do
+    query getStudentAssignmentsQuery (student, course)
+  where
+    getStudentAssignmentsQuery :: Query
+    getStudentAssignmentsQuery = [q|
+        select     course_id, student_addr, desc
+        from       StudentAssignments
+        left join  Assignments
+               on  assignment_hash = Assignments.hash
+        where      student_addr    = ?
+               and course_id       = ?
+    |]
+
+submitAssignment :: DBM m => SignedSubmission -> m (Hash Submission)
+submitAssignment = createSignedSubmission
+
+createSignedSubmission :: DBM m => SignedSubmission -> m (Hash Submission)
+createSignedSubmission sigSub = do
+    let
+        submission     = sigSub^.ssSubmission
+        submissionSig  = sigSub^.ssWitness.swSig
+
+        student        = submission^.sStudentId
+        submissionHash = submission^.to hash
+        submissionType = submission^.sType
+        assignment     = submission^.sAssignment
+
+        course         = assignment^.aCourseId
+        assignmentHash = assignment^.to hash
+
+    _ <- existsStudent student        `assert`     StudentDoesNotExist    student
+    _ <- existsCourse  course         `assert`     CourseDoesNotExist     course
+    _ <- getAssignment assignmentHash `assertJust` AssignmentDoesNotExist assignmentHash
+
+    execute setAssignmentAuthorRequest (student, assignmentHash)
```

Comment:
Ok. I'll add the ability to do transactions to the `MonadSQLiteDB` class, then.

Oh, wait, `sqlite-simple` cannot into transactions - it will run only first command and call it a day.
I'll use methods from `direct-sqlite` then.

Reply:
@Heimdell I think you're wrong. https://hackage.haskell.org/package/sqlite-simple-0.4.16.0/docs/Database-SQLite-Simple.html#v:withTransaction

Reply:
I tried to do transactions in SQL directly, like `begin transaction; ...`. Thanks for pointing out this method.